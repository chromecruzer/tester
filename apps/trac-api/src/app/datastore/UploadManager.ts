import util from "util";
import _ from "lodash";
import path from "path";
import {DateTime} from "luxon";
import {UploadConfiguration} from "../ApiConfiguration";
import {
  dateNow,
  supportedUploadFiles,
  UploadMetadata,
  UploadSupportedFileTypes
} from "@trac/datatypes";

const userSessionsFile = 'activeSessions.json';
const uploadMetadataFile = 'uploadFileProperties.json';
import DataStoreUtilities from "./DataStoreUtilities";


const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});


export interface UserSession {
  token: string;
  uploadName: string;
  type: UploadSupportedFileTypes;
  user: string;
  created: Date;
  form: { [id: string]: string };
}

export default class UploadManager {
  public sessions: { [id: string]: UserSession } = {};
  public metadata: UploadMetadata[] = [];
  public uploadPath: string;
  private readonly filePath: (filename) => string;
  private dsu: DataStoreUtilities;

  constructor(private config: UploadConfiguration) {
    this.uploadPath = path.join(...this.config.root);
    this.filePath = filename => path.join(this.uploadPath, filename);
    this.dsu = new DataStoreUtilities();
  }

  public async initialize() {
    const err = [];
    await this.checkUploadDirectories()
      .catch(e => err.push(...e));
    await this.checkFolder('audit_scans');
    await this.loadJson(userSessionsFile)
      .catch(e => console.info('no sessions found'));
    Object.keys(this.sessions).map(async s => {
      const session = this.sessions[s];
      delete this.sessions[s]
      return this.dsu.remove(session.uploadName)
        .catch(e => err.push(new Error(`Cannot remove ${session.uploadName} because ${e.message}`)));
    });
    await this.saveJson(userSessionsFile)
      .catch(e => err.push(e));
    await this.loadJson(uploadMetadataFile)
      .catch(e => console.log('upload archive is empty'));
    if (err.length > 0) {
      return Promise.reject(err);
    }
    console.log('initialize is done');
    // console.log(`archives ${dump(this.metadata)}`)
  }

  private async checkFolder(folder, err = []) {
    const dir = path.join(this.uploadPath, folder);
    console.log(`checking upload folder ${dir}`);
    return this.dsu.mkdir(dir)
      .catch(() =>
        err.push(new Error(`Cannot create upload folder "${folder}" in ${this.uploadPath}`)));
  }

  private async checkUploadDirectories() {
    const err = [];
    supportedUploadFiles().forEach(f => this.checkFolder(f,err));
    if (err.length > 0) {
      return Promise.reject(err);
    }
  }

  public async removeSession(token: string, removeFile = false) {
    const err = [];
    const session = this.sessions[token];
    if (session) {
      delete this.sessions[token];
      if (removeFile) {
        await this.dsu.remove(session.uploadName)
          .catch(e => err.push(new Error(`Cannot remove ${session.uploadName} because ${e.message}`)));
      }
      await this.saveJson(userSessionsFile)
        .catch(e => err.push(e));
      if (err.length > 0) {
        return Promise.reject(err);
      }
    }
  }

  public async addSession(filepath, user, type, form) {
    let err = null;
    const result = {
      token: UploadManager.newToken(),
      type,
      uploadName: filepath,
      user,
      created: DateTime.now().toJSDate(),
      form,
    };
    this.sessions[result.token] = result;
    // console.log(`sessions ${dump(this.sessions)}`)
    await this.saveJson(userSessionsFile)
      .catch(e => err = e);
    if (err) {
      return Promise.reject(err);
    }
    return result.token;
  }

  public async archiveSession(token: string) {
    const err = [];
    const session = this.sessions[token] || null;
    if (session === null) {
      return Promise.reject([new Error(`Session with token ${token} does not exist`)]);
    }
    // console.log(`archiving session ${dump(session)}`);
    const archived = {
      name: session.uploadName,
      type: session.type,
      user: session.user,
      date: session.created,
      form: session.form
    };
    this.metadata.push(archived);
    await this.removeSession(token)
      .catch(e => err.push(...e));
    await this.saveJson(uploadMetadataFile)
      .catch(e => err.push(e));
    if (err.length > 0) {
      return Promise.reject(err);
    }
  }

  public async markRefresh(user) {
    const err = [];
    const archived = {
      type: 'refresh_data' as UploadSupportedFileTypes,
      user,
      date: dateNow(),
    };
    this.metadata.push(archived);
    await this.saveJson(uploadMetadataFile)
      .catch(e => err.push(e));
    if (err.length > 0) {
      return Promise.reject(err);
    }
  }

  public history() {
    // console.log('setting up history');
    return (req, res) => {
      // console.log('returning history')
      res.json(_.filter(this.metadata, m => m.type !== 'audit_scans'));
    }
  }

  private async loadJson(filename: string) {
    const filepath = this.filePath(filename);
    // console.log(`checking path ${filepath}`);
    if (!await this.dsu.check(filepath)) {
      return;
    }
    // console.log(`reading from path ${filepath}`);
    return this.dsu.retrieve(filepath)
      .then(json => {
        // console.log(`retrieved ${dump(json)}`);
        if (filename === userSessionsFile) {
          this.sessions = json;
        } else {
          this.metadata = json;
        }
      })
      .catch(e => new Error(`File ${filepath} could not be read. ${e.message}`));
  }

  private async saveJson(filename) {
    const filepath = this.filePath(filename);
    const json = filename === userSessionsFile ? this.sessions : this.metadata;
    // console.log(`writing ${dump(json)} to ${filepath}`);
    return this.dsu.store(filepath, json)
  }

  static newToken() {
    return Math.random().toString(36).substring(0, 10)
      .replace('.', '');
  }
}
