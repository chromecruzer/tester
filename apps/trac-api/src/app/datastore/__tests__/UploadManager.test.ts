import * as path from "path";
import UploadManager from "../UploadManager";
import * as util from "util";
import {DateTime} from "luxon";

const dump = (obj, depth = null) => util.inspect(obj, {depth: depth, colors: true});

describe('UploadManager', () => {
  let manager, config, mockDsu, rootPath;
  const createMockSession = () => {
    const expected = {token: 'new token', date: DateTime.fromObject({day:10, month: 10, year: 2020})};
    UploadManager.newToken = jest.fn().mockReturnValue(expected.token);
    Date.now = jest.fn().mockReturnValue(expected.date);
    return expected;
  };
  beforeEach(() => {
    config = {root: ['some', 'root', 'directory']};
    rootPath = path.join(...config.root, 'iol_report');
    mockDsu = {
      mkdir: jest.fn().mockReturnValue(Promise.resolve()),
      check: jest.fn().mockReturnValue(Promise.resolve(true)),
      remove: jest.fn().mockReturnValue(Promise.resolve()),
      store: jest.fn().mockReturnValue(Promise.resolve()),
      retrieve: jest.fn().mockReturnValue(Promise.resolve({}))
    };
    manager = new UploadManager(config);
    manager.dsu = mockDsu;
  });
  describe('initialize', () => {
    it('should check that the upload directory exists', () => {
      return manager.initialize().then(() => {
        expect(mockDsu.mkdir).toHaveBeenCalledWith(rootPath);
      });
    });
    it('should remove any sessions that exist', () => {
      const [filepath1, filepath2] = ['file1', 'file2'];
      mockDsu.retrieve.mockReturnValue(Promise.resolve({
        token1: {
          token: 'token1',
          type: 'IolReport',
          user: 'uploadUser',
          uploadName: filepath1,
          created: Date.now(),
        },
        token2: {
          token: 'token2',
          type: 'IolReport',
          user: 'uploadUser',
          uploadName: filepath2,
          created: Date.now(),
        }
      }));
      return manager.initialize().then(() => {
        expect(mockDsu.remove).toHaveBeenCalledWith(filepath1);
        expect(mockDsu.remove).toHaveBeenCalledWith(filepath2);
        expect(mockDsu.store)
          .toHaveBeenCalledWith(path.join(...config.root, 'activeSessions.json'), {});
      });
    });
    it('should handle fs.mkdir errors', () => {
      mockDsu.mkdir.mockReturnValue(Promise.reject("Rejecting mkdir"));
      return expect(manager.initialize()).rejects.toMatchSnapshot();
    });
    it('should handle missing sessions file', () => {
      const expected = ['this', 'is', 'an', 'array'];
      mockDsu.retrieve.mockImplementationOnce(() => Promise.resolve(expected));
      mockDsu.check.mockImplementationOnce(() => Promise.resolve(false));
      mockDsu.check.mockImplementationOnce(() => Promise.resolve(true));
      return manager.initialize()
        .then(() => {
          expect(manager.sessions).toEqual({});
          expect(manager.metadata).toEqual(expected);
        });
    });
    it('should handle missing archive file', () => {
      const expected = {this: 'is', a: 'json', file: '!'};
      mockDsu.retrieve.mockImplementationOnce(() => Promise.resolve(expected));
      mockDsu.check.mockImplementationOnce(() => Promise.resolve(true));
      mockDsu.check.mockImplementationOnce(() => Promise.resolve(false));
      return manager.initialize()
        .then(() => {
          expect(manager.sessions).toEqual(expected);
          expect(manager.metadata).toEqual([]);
        });
    });
  });
  describe('removeSession', () => {
    let originalSessions;
    beforeEach(() => {
       originalSessions = {
        'to-be-removed': {
          token: 'to-be-removed',
          type: 'IolReport',
          uploadName: '/some/uploaded/file',
          user: 'sirbilliesbe',
          created: 'a beautiful date'
        },
        'to-keep': {
          token: 'to-keep',
          type: 'IolReport',
          uploadName: '/some/uploaded/file/2b/archived',
          user: 'sirbilliesbe',
          created: 'a rainy date'
        }
      };
      manager.sessions = {...originalSessions};
    });
    it('should remove a session fron the userSessionsFile', () => {
      return manager.removeSession('to-be-removed')
        .then(() => {
          const actualSessions = mockDsu.store.mock.calls[0][1];
          expect(actualSessions).toHaveProperty('to-keep', originalSessions['to-keep']);
          expect(Object.keys(actualSessions)).not.toContain('to-be-removed');
        });
    });
    it('should should remove the uploaded file of the session when requested', () => {
      return manager.removeSession('to-be-removed', true)
        .then(() => {
          expect(mockDsu.remove).toBeCalledWith(originalSessions['to-be-removed'].uploadName);
        });
    });
    it('should should handle an error from remove', () => {
      mockDsu.remove.mockImplementationOnce(() => Promise.reject(new Error("Rejecting fs.remove")));
      return expect(manager.removeSession('to-be-removed', true)).rejects.toMatchSnapshot();
    });
  });
  describe('addSession', () => {
    it('should add a session from the userSessionsFile', () => {
      const expected = createMockSession();
      const [expectedFile, expectedUser, expectType] = ['/some/file', 'jqpublic', 'IolReport'];
      return manager.addSession(expectedFile, expectedUser, expectType)
        .then(actualToken => {
          expect(actualToken).toEqual(expected.token);
          const actual = mockDsu.store.mock.calls[0][1];
          // console.log(`actual session ${dump(actual)}`);
          expect(actual).toHaveProperty(actualToken);
          expect(actual[actualToken]).toHaveProperty('uploadName', expectedFile);
          expect(actual[actualToken]).toHaveProperty('user', expectedUser);
          expect(actual[actualToken]).toHaveProperty('type', expectType);
          expect(actual[actualToken]).toHaveProperty('created', expected.date.toJSDate());
          expect(actual[actualToken]).toHaveProperty('token', expected.token);
        });
    });
    it('should should handle an error from saveJson', () => {
      mockDsu.store.mockReturnValue(Promise.reject(new Error("Rejecting store")));
      const expected = createMockSession();
      const [expectedFile, expectedUser, expectType] = ['/some/file', 'jqpublic', 'IolReport'];
      return expect(manager.addSession(expectedFile, expectedUser, expectType)).rejects.toMatchSnapshot();
    });
  });
  describe('archiveSession', () => {
    let originalSessions;
    beforeEach(() => {
      originalSessions = {
        'to-be-archived': {
          token: 'to-be-archived',
          type: 'IolReport',
          uploadName: '/some/uploaded/file/2b/archived',
          user: 'sirbilliesbe',
          created: 'a beautiful date'
        },
        'to-keep': {
          token: 'to-keep',
          type: 'IolReport',
          uploadName: '/some/uploaded/file',
          user: 'sirbilliesbe',
          created: 'a rainy date'
        }
      };
      manager.sessions = {...originalSessions};
    });
    it('should transfer a session from user sessions to the archive and add the uuid object', () => {
      const archiveToken = 'to-be-archived';
      return manager.archiveSession(archiveToken)
        .then(() => {
          const writtenSessions = mockDsu.store.mock.calls[0][1];
          const writtenArchives = mockDsu.store.mock.calls[1][1];
          expect(writtenSessions).not.toHaveProperty(archiveToken);
          expect(writtenArchives).toMatchSnapshot();
        });
    });
    it('should should handle an error from remove session', () => {
      mockDsu.store.mockImplementationOnce(() => Promise.reject(new Error("Rejecting store")));
      return expect(manager.archiveSession('to-be-archived')).rejects.toMatchSnapshot();
    });
    it('should should handle an error from saveJson', () => {
      mockDsu.store.mockImplementationOnce(() => Promise.resolve());
      mockDsu.store.mockImplementationOnce(() => Promise.reject(new Error("Rejecting store")));
      return expect(manager.archiveSession('to-be-archived')).rejects.toMatchSnapshot();
    });
  });
});
