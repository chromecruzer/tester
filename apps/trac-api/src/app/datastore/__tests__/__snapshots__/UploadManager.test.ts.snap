// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadManager addSession should should handle an error from saveJson 1`] = `[Error: Rejecting store]`;

exports[`UploadManager archiveSession should should handle an error from remove session 1`] = `
Array [
  [Error: Rejecting store],
]
`;

exports[`UploadManager archiveSession should should handle an error from saveJson 1`] = `
Array [
  [Error: Rejecting store],
]
`;

exports[`UploadManager archiveSession should transfer a session from user sessions to the archive and add the uuid object 1`] = `
Array [
  Object {
    "date": "a beautiful date",
    "form": undefined,
    "name": "/some/uploaded/file/2b/archived",
    "type": "IolReport",
    "user": "sirbilliesbe",
  },
]
`;

exports[`UploadManager initialize should handle fs.mkdir errors 1`] = `
Array [
  [Error: Cannot create upload folder "audit_scans" in some/root/directory],
  [Error: Cannot create upload folder "iol_report" in some/root/directory],
  [Error: Cannot create upload folder "customer_data" in some/root/directory],
  [Error: Cannot create upload folder "customer_contact_data" in some/root/directory],
  [Error: Cannot create upload folder "product_data" in some/root/directory],
  [Error: Cannot create upload folder "sales_mapping_data" in some/root/directory],
  [Error: Cannot create upload folder "sales_rep_data" in some/root/directory],
]
`;

exports[`UploadManager removeSession should should handle an error from remove 1`] = `
Array [
  [Error: Cannot remove /some/uploaded/file because Rejecting fs.remove],
]
`;
